cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -283,51 +236,62 @@ public void setStop(final Id.Program id, final String pid, final long endTime, f
   public void setStop(final Id.Program id, final String pid, final long endTime, final ProgramRunStatus runStatus,
                       final BasicThrowable failureCause) {
     Preconditions.checkArgument(runStatus != null, "Run state of program run should be defined");
-    appsTx.get().executeUnchecked(new TransactionExecutor.Function<AppMetadataStore, Void>() {
+    txExecute(transactional, new TxRunnable() {
       @Override
-      public Void apply(AppMetadataStore mds) throws Exception {
-        mds.recordProgramStop(id, pid, endTime, runStatus, failureCause);
-        return null;
-      }
-    }, apps.get());
+      public void run(DatasetContext context) throws Exception {
+        AppMetadataStore metaStore = getAppMetadataStore(context);
+        metaStore.recordProgramStop(id, pid, endTime, runStatus, failureCause);
+
+        // This block has been added so that completed workflow runs can be logged to the workflow dataset
+        if (id.getType() == ProgramType.WORKFLOW && runStatus == ProgramRunStatus.COMPLETED) {
+          RunRecordMeta runRecordMeta = metaStore.getRun(id, pid);
+          if (runRecordMeta == null) {
+            return;
+          }
 
-    // This block has been added so that completed workflow runs can be logged to the workflow dataset
-    if (id.getType() == ProgramType.WORKFLOW && runStatus == ProgramRunStatus.COMPLETED) {
-      Id.Workflow workflow = Id.Workflow.from(id.getApplication(), id.getId());
-      recordCompletedWorkflow(workflow, pid);
-    }
-    // todo: delete old history data
+          recordCompletedWorkflow(metaStore, getWorkflowDataset(context),
+                                  Id.Workflow.from(id.getApplication(), id.getId()), pid);
+        }
+        // todo: delete old history data
+      }
+    });
   }
 
-  private void recordCompletedWorkflow(final Id.Workflow id, String pid) {
-    final RunRecordMeta run = getRun(id, pid);
-    if (run == null) {
+  private void recordCompletedWorkflow(AppMetadataStore metaStore, WorkflowDataset workflowDataset,
+                                       Id.Workflow workflowId, String runId) {
+    RunRecordMeta runRecord = metaStore.getRun(workflowId, runId);
+    if (runRecord == null) {
       return;
     }
-    Id.Application app = id.getApplication();
-    ApplicationSpecification appSpec = getApplication(app);
-    if (appSpec == null || appSpec.getWorkflows() == null || appSpec.getWorkflows().get(id.getId()) == null) {
+    Id.Application app = workflowId.getApplication();
+    ApplicationSpecification appSpec = getApplicationSpec(metaStore, workflowId.getApplication());
+    if (appSpec == null || appSpec.getWorkflows() == null || appSpec.getWorkflows().get(workflowId.getId()) == null) {
       return;
```

Comment:
It think it is possible since we didn't check if the app spec is still there when the `setStop()` method is called. Anyway, I will add a warn there.